cmd/sfncli/runner.go

```diff
@@ -20,99 +25,170 @@ const maxTaskOutputLength = 32768
+				}
+				return t.sendTaskFailure(TaskFailureCommandExitedNonzero{stderr: stderr})
+			case status.Signaled() && status.Signal() == syscall.SIGKILL:
+				return t.sendTaskFailure(TaskFailureCommandKilled{stderr: stderr})
+			}
+		}
+		return t.sendTaskFailure(TaskFailureUnknown{err})
+	}
+	t.logger.Info("exec-command-end")
 
-	log.InfoD("exec-command-start", logger.M{
-		"args": args, "cmd": t.cmd, "execution_name": executionName,
-	})
-	if err := cmd.Run(); err != nil {
-		return t.handleProcessError(ctx, "exec-command-err", executionName, err)
+	// AWS / states language requires JSON output
+	taskOutput := taskOutputFromStdout(stdoutbuf.String())
+	var taskOutputMap map[string]interface{}
+	if err := json.Unmarshal([]byte(taskOutput), &taskOutputMap); err != nil {
+		return t.sendTaskFailure(TaskFailureTaskOutputNotJSON{output: taskOutput})
 	}
-	log.InfoD("exec-command-end", logger.M{"execution_name": executionName})
-
-	// AWS requires JSON output. If it isn't, then make it so.
-	output := stdoutbuf.String()
-	var out map[string]interface{}
-	if err := json.Unmarshal([]byte(output), &out); err != nil {
-		wrappedErr := fmt.Errorf("Worker must output json object to stdout: %s", err)
-		return t.handleProcessError(ctx, "process-output", executionName, wrappedErr)
+	if executionName != nil {
+		taskOutputMap["_EXECUTION_NAME"] = *executionName
 	}
-	out["_EXECUTION_NAME"] = executionName
 
-	marshaledOut, err := json.Marshal(out)
+	finalTaskOutput, err := json.Marshal(taskOutputMap)
 	if err != nil {
-		return t.handleProcessError(ctx, "process-output", executionName, err)
+		return t.sendTaskFailure(TaskFailureUnknown{fmt.Errorf("JSON output re-marshalling failed. This should never happen. %s", err)})
 	}
 	_, err = t.sfnapi.SendTaskSuccessWithContext(ctx, &sfn.SendTaskSuccessInput{
-		Output:    aws.String(string(marshaledOut)),
+		Output:    aws.String(string(finalTaskOutput)),
 		TaskToken: &t.taskToken,
 	})
 	if err != nil {
-		log.ErrorD(
-			"process-success", logger.M{"error": err.Error(), "execution_name": executionName},
-		)
+		t.logger.ErrorD("send-task-success-error", logger.M{"error": err.Error()})
 	}
 
 	return err
 }
+
+func (t *TaskRunner) handleSignals(ctx context.Context) {
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan)
+	defer signal.Stop(sigChan)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case sigReceived := <-sigChan:
+			if t.execCmd.Process == nil {
+				continue
+			}
+			pid := t.execCmd.Process.Pid
+			// SIGTERM is special. If it gets sent to sfncli, initiate a docker-stop like shutdown process:
+			// - forward the SIGTERM to the command
+			// - after a grace period (5s) send SIGKILL to the command if it's still running
+			if sigReceived == syscall.SIGTERM {
+				t.receivedSigterm = true
+				go func(pidtokill int) {
+					time.Sleep(5 * time.Second) // grace period
+					signalProcess(pidtokill, os.Signal(syscall.SIGKILL))
```

Comment:
Shouldn't SIGTERM/SIGKILL also cause the sfncli process to exit?

Reply:
hmm yes... SIGKILL will automatically kill it, but I should add logic in main() to stop the polling loop on SIGTERM

Reply:
nvm... there's already a signal handler that does this in main()

just tested this as well: sent `sfncli` SIGTERM while it was running a command, and it sent the correct error message and shutdown: https://us-west-2.console.aws.amazon.com/states/home?region=us-west-2#/executions/details/arn:aws:states:us-west-2:589690932525:execution:raf--cil-reliability-dashboard-master--4--development:61e6bbeb-2459-4710-8c02-3691dce1068b

sent